internal/authentication/http.go

```diff
@@ -0,0 +1,237 @@
+package authentication
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/go-chi/chi"
+	"golang.org/x/oauth2"
+)
+
+// contextKey to use when setting context values in the HTTP package.
+type contextKey string
+
+// String implements the Stringer interface and makes it
+// nice to print contexts.
+func (c contextKey) String() string {
+	return "HTTP context key " + string(c)
+}
+
+const (
+	state = "I love Observatorium"
+	// tenantKey is the key that holds the tenant in a request context.
+	tenantKey contextKey = "tenant"
+)
+
+// WithTenant finds the tenant from the URL parameters and adds it to the request context.
+func WithTenant(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		tenant := chi.URLParam(r, "tenant")
+		next.ServeHTTP(w, r.WithContext(
+			context.WithValue(r.Context(), tenantKey, tenant),
+		))
+	})
+}
+
+// NewTenantHeaderMiddleware returns a middleware that adds the ID of the tenant to the specified header.
+func NewTenantHeaderMiddleware(header string, tenantIDs map[string]string) Middleware {
```

Comment:
If we inline this with `r.Use()` then I would prefer `r.Use(WithTenantHeader(headerFlag, IDs))`. Maybe we can rename to that style. See `WithTenant` above.